File: examples/ts/runner.ts
// Example Gota runner (TypeScript): FNV-1a over the buffer. See ../README.md.
//   node --experimental-strip-types runner.ts [buffer_bytes] [warmup_s] [measure_s]
//
// JS has no native u64, so the 64-bit FNV runs on BigInt. That is genuinely slow
// (the result is honest, not a bug): it shows the cost of naive 64-bit math in JS,
// which the example intro already flags as not a serious cross-language comparison.
import { run, type Bencher } from "./gota.ts";

const PRIME = 0x100000001b3n;
const MASK = 0xffffffffffffffffn;

run("ts", (b: Bencher, data: Uint8Array) => {
  b.bench("fnv1a-64", () => {
    let h = 0xcbf29ce484222325n;
    for (let i = 0; i < data.length; i++) {
      h = ((h ^ BigInt(data[i])) * PRIME) & MASK;
    }
    data[0] ^= Number(h & 0xffn); // sink: consume h
  });
});

File: examples/cpp/runner.cpp
// Example Gota runner (C++): FNV-1a over the buffer. See ../README.md.
#include "gota.hpp"

#include <cstdint>
#include <span>
#include <string_view>

namespace {
constexpr std::string_view IMPL = "cpp";
}

int main(int argc, char** argv) {
    return gota::run(IMPL, argc, argv,
                     [](gota::Bencher& b, std::span<std::uint8_t> data) {
        b.bench("fnv1a-64", [&] {
            std::uint64_t h = 0xcbf29ce484222325ULL;
            for (const std::uint8_t x : data) {
                h = (h ^ x) * 0x100000001b3ULL;
            }
            data[0] ^= static_cast<std::uint8_t>(h);  // sink: consume h
        });
    });
}
